### DISCUSSION TITLE
installation of runtime packages

### DISCUSSION THREAD

**Discussion by meirumeiru** [microsoft/WindowsAppSDK#2099] [Q&A]
Let's talk about "AppX packages" ... or MSIX files.

As far as I know they come from the "store world" and therefore were built to support "per user" and "in sandbox" installations. But today Microsoft allows us to use e.g. WinUI3 even for unpackaged apps. And here comes my question.

Assume I have an application with an MSI (not MSIX) or EXE setup. And this setup installs an application for all users. What do I have to do with the "Windows App Runtime" ??

Internally the setup provided for the framework seems to do this:

1) calls PackageManager::AddPackageAsync -> that's the part I understand... almost... my question would be... is this then installed "to the user" or "to the machine" and ... what does it mean to install "per user"? ... like writing COM-registrations to HKCU isntead of HKLM?

2) if it fails, it calls PackageManager::RegisterPackageByFullNameAsync ... ok... this is something I completely don't understand... does this mean, that I can only use the package, when I did run the setup for every use that's using my app and thus this package? ... so... is this then not a "per machine" installation???

3) then it "provisions" the app package... but not, if it's a "framework"... even more questions... why not for frameworks? because frameworks are provisioned always? if it's not possible and they have to be installed on a per user basis? if... no idea why... ... and, if provisioning is not possible for "framework"s, then, how could the step 2 (register) work in case it is needed??

... sorry, I don't get it. I'd like to understand all the "behind" stuff... I don't feel good, if I'm delivering software the does things I don't fully understand.

**Comment by ghost** [comment]
> Assume I have an application with an MSI (not MSIX) or EXE setup. And this setup installs an application for all users. What do I have to do with the "Windows App Runtime" ??

WindowsAppRuntime is an MSIX package  therefore **Any App that heavily uses/solely depends on Windows App SDK components (such as WinUI 3)** can't be installed  _"per machine"_ because to this date,  MSIX shite is/are always installed/registered per user basis. >_<   (you can ask msft's @DrusTheAxe  for  clarification.)



provisioned means the package will be installed/registered for each users _separately_ (still per users) , unlike "per machine" where all users just share a common exe.

**Comment by DrusTheAxe** [comment]
> Assume I have an application with an MSI (not MSIX) or EXE setup. And this setup installs an application for all users. What do I have to do with the "Windows App Runtime" ??

You'd want to add _and provision_ the packages (and install any licenses).

The trivial answer, "Run WindowsAppRuntineInstaller.exe elevated". The installer does all the right things to get WinAppSDK on the box for the current user. AND, if it's run elevated, it _ALSO_ provisions<sup>1</sup> the packages for all users.

This is effectively:

```
FOREACH p IN *.msix
    packagemanager.AddPackageAsync(p)
    IF !p.IsFramework and IsElevated()
        packagemanager.ProvisionPackagesForAllUsersAsync(p.packagefamilyname)
BootstrapInitialize()
h = LoadLibrary("Microsoft.WindowsAppRuntime.dll")
f = GetProcAddress("MsixInstallLicenses")
f()
```

That's functionally what the installer does - install the packages, install their licenses (if any) and, if running elevated, provision the packages.

<sup>1</sup> "Provisioning" is sometimes referred to as "preinstalled". Same thing.

**Comment by DrusTheAxe** [comment]
> 1. is this then installed "to the user"

yes

>what does it mean to install "per user"?

It means the package is available for use by a user, and updating or removing a package for one user does not affect other users. 

>like writing COM-registrations to HKCU isntead of HKLM?

Like that, yes. Though the implementation details vary.

That's how you'd install a classic COM object for 1 user but not others, and how you could change or remove the COM object for 1 user without affecting others.

MSIX does the equivalent thing, but the implementation varies.
* For some data it means scribbling in HKCU instead of HKLM
* For other data, it means scribbling the data in a common / shared location (HKLM and/or elsewhere) with tags indicating which user(s) have access to it, and later accessing that data do respects that

That's why 5 users running `Add-AppxPackage LolzCatzVideoz.msix` doesn't take 5x the disk space. First user causes the package to be laid down on disk and ACL'd (aka "staged") so it's readable/executable but not writable (except by Windows). Second+ users don't need another physical copy, the bits are the same for all users and can't be altered so everyone can share the same readonly copy. Windows tracks which users have the package registered so they can use (and others don't).

That's also why when that 5th user runs Remove-AppxPackage you don't see huge disk savings -- because that package needs to stay on disk for those other 4 users that have it registered. It's when the _last_ user deregisters the package and it's not otherwise needed we can remove the package (without breaking anyone). This is managed by using a common / shared location and tracking which users have what package.

Exactly which bits are physically stored per-user vs which are physically stored in a shared location and managed per-user depends on the particular bits. Regardless, the end experience is the same -- one user can and remove packages without negatively impact other users. That's why MSIX is said to support a "per-user" model.

**Comment by DrusTheAxe** [comment]
>2. if it fails, it calls PackageManager::RegisterPackageByFullNameAsync...

It's common to say you "install" a package but technically that never happens. There is no 'install' per se. The deployment pipeline is geared around 3 fundamental concepts

* Index
* Stage
* Register

The (grossly oversimplified) explanation:
1. Index -- parse and store the package's information (apps defined, path where it's located, etc) so Windows knows and can reason about the package
2. Stage -- MD pkgdir + UNZIP msix into pkgdir + ACL pkgdir
3. Register -- write up the package to a user's profile

1+2 are 1 time operations. No matter how many users have a packaged installed, we only need to do these once on a machine. This just puts some bits on the box; they're not usable yet. They're inert.

3 is a per-user operation. This wires up the package to your profile - add tiles to StartMenu, associate .txt with apps, etc. This is what makes the package's content usable.

`PackageManager.AddPackage*()` does 1-3 and the common way it's handled.
`PackageManager.StagePackage*()` does 1-2 and `.RegisterPackage()` does 3.

"Provisioning a package" means "add the package to the list of provisioned packages". Windows knows all users should have all provisioned packages so e.g. at logon if you don't have some they're registered for you.

If you're an installer (MSIX, EXE, whatever) running as LocalSystem and you want to add packages on the machine for all users instead of

```
PackageManager.AddPackage*()
```

you'd want to call

```
PackageManager.StagePackage*()`
PackageManager.ProvisionPackageForAllUsers()
```

That will Index+Stage the package (i.e. add it onto the machine) and add it to the provisioned list, so all users get it.

NOTE: MSIX currently doesn't support registering a package for `LocalSystem`. That's why if you're installer is running as LocalSystem you want to Stage*() + Provision*() rather than Add*().

So that's why `PackageManager` has no `InstallPackage()` API, it's `Add*()`, `Stage*()` and `Register*()`.

We commonly say "install" a package but it's an informal simplification. Technically the package is "indexed, staged and registered". But that's a mouthful and somewhat pedantic for most folks. "Install a package" conveys the essence.

Unless you're writing an installer. Or like work where the distinction is important.

**Comment by DrusTheAxe** [comment]
> then it "provisions" the app package... but not, if it's a "framework"... even more questions... why not for frameworks?

Because Windows doesn't support provisioning Framework packages (as in, PackageType=Framework).

You can provision Main packages, and if the Main package has dependencies on framework packages that's tracked (so we know we need framework package X and don't remove it and thus break the next time we register that provisioned package for some user). But we don't _directly_ track framework packages in the provisioning list.

Or at least, we didn't. We might now in Win11 (I don't recall). But Windows App SDK supports down to [RS5 (10.0.17763.0)](https://en.wikipedia.org/wiki/Windows_10_version_history#Channels) so provisioning all but the framework package gets the same net result.
